### PR TITLE
Make flatfs robust

### PR DESCRIPTION
**Pull request by Stebalien**
1. Improve query logic to finally support all features. Offsets, limits, filters, and sorting are all supported through naive queries.
2. Fetch files from disk when asked. Not exactly fast but the user is allowed to dig their own grave.
3. Explicitly restrict keys instead of randomly doing bad things at runtime.

Importantly, this change only allows keys of the form `/[0-9A-Z+-_=]`. That is, upper-case alphanumeric keys in the root namespace (plus some special characters).

Why? We don't encode keys before writing them to the filesystem. This change ensures that:

1. Case sensitivity doesn't matter because we only allow upper-case keys. 
2. Path separators and special characters doesn't matter.

For context, go-ipfs only uses flatfs for storing blocks. Every block CID is encoded as uppercase alphanumeric text (specifically, uppercase base32).

We could be less restrictive, but this is safer and easier to understand. Unfortunately, we _can't_ allow mixed case (Windows) and can't allow lowercase because we're already using uppercase keys.

---

fixes #32, #61, #40, #22

tests with the query test suite

### CODE REVIEW

File: flatfs.go [hunk 1]
```diff
@@ -1042,11 +1071,30 @@ func (fs *Datastore) walk(path string, result *query.ResultBuilder) error {
...
 		}
 
+		var (
+			value []byte
+			size  int
+			err   error
+		)
+		if !result.Query.KeysOnly {
+			value, err = ioutil.ReadFile(filepath.Join(path, fn))
+			size = len(value)
```
Comment (dirkmc) [inline flatfs.go]: Is it safe to assume that if there's an error, value will have length 0? (seems like it should be, just want to confirm)

File: key_test.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	"testing"
+
+	"github.com/ipfs/go-datastore"
+)
+
+var (
+	validKeys = []string{
+		"/FOO",
+		"/1BAR1",
+		"/=EMACS-IS-KING=",
```
Comment (hsanjuan) [inline key_test.go]: :D

File: flatfs.go [hunk 3]
```diff
@@ -1090,12 +1138,17 @@ func (fs *Datastore) Batch() (datastore.Batch, error) {
 }
 
 func (bt *flatfsBatch) Put(key datastore.Key, val []byte) error {
+	if !keyIsValid(key) {
+		return fmt.Errorf("key not supported by flatfs: '%q'", key)
```
Comment (hsanjuan) [inline flatfs.go]: Nitpick: Might add a new `ErrInvalidKey` to the top and wrap it.

File: flatfs.go [hunk 4]
```diff
@@ -682,7 +700,18 @@ func (fs *Datastore) Query(q query.Query) (query.Results, error) {
...
+	results := b.Results()
+
+	if len(q.Filters) > 0 ||
+		len(q.Orders) > 0 ||
+		q.Limit > 0 ||
+		q.Offset > 0 {
+		nq := q
+		nq.Prefix = "" // already handled
+		results = query.NaiveQueryApply(nq, results)
+	}
```
Comment (hsanjuan) [inline flatfs.go]: So `NaiveQueryApply` will already do nothing if it does not have to do anything so you might skip the checks here as redundant: https://github.com/ipfs/go-datastore/blob/master/query/query_impl.go#L119

File: flatfs.go [hunk 5]
```diff
@@ -654,17 +678,11 @@ func (fs *Datastore) doDelete(key datastore.Key) error {
...
-		!q.KeysOnly ||
-		q.ReturnExpirations ||
-		q.ReturnsSizes {
-		// TODO this is overly simplistic, but the only caller is
-		// `ipfs refs local` for now, and this gets us moving.
-		return nil, errors.New("flatfs only supports listing all keys in random order")
+	if prefix != "/" {
+		// This datastore can't include keys with multiple components.
+		// Therefore, it's always correct to return an empty result when
+		// the user requests a filter by prefix.
```
Comment (hsanjuan) [inline flatfs.go]: Log a warning message? Very likely indicates bad usage of the datastore which might happen silently to the application above.

File: key_test.go (hunk 2 above)
Comment (Stebalien) [reply key_test.go]: Clearly.

File: flatfs.go (hunk 1 above)
Comment (Stebalien) [reply flatfs.go]: Hm. Good question. No. It'll return the partial value and an error.  I'll change this.

File: flatfs.go (hunk 4 above)
Comment (Stebalien) [reply flatfs.go]: Ah, well, https://github.com/ipfs/go-datastore/pull/147 (now merged).

Comment (hsanjuan) [reply flatfs.go]: You are setting prefix to `""` anyways so it would have not entered the `if`

**Comment by Stebalien** [comment]
> Any relevant context/request that makes this a necessary change or is it just general maintenance?

General maintenance. I needed to pull through a fix for prefix handling in the query logic.

**Comment by hsanjuan** [comment]
Errors in tests, seems circle is building with old Go ?

**Comment by Stebalien** [comment]
> Errors in tests, seems circle is building with old Go ?

Updated go-log.

**Comment by Stebalien** [comment]
Oh, and probably old go.

**Pull request merged**